csrc/tma.cpp

```diff
@@ -0,0 +1,391 @@
+
+  // Checks based on the documentation of cuTensorMapEncodeTiled, error messages
+  // are mostly directly copied from the doc
+
+  TORCH_INTERNAL_ASSERT(
+      tensor_rank != 0 && tensor_rank <= 5,
+      "tensorRank must be non-zero and less than or equal to the maximum supported dimensionality of 5");
+  if (interleave != CU_TENSOR_MAP_INTERLEAVE_NONE) {
+    TORCH_INTERNAL_ASSERT(
+        tensor_rank >= 3,
+        "If interleave is not CU_TENSOR_MAP_INTERLEAVE_NONE, then tensorRank must additionally be greater than or equal to 3.");
+  }
+
+  if (interleave == CU_TENSOR_MAP_INTERLEAVE_32B) {
+    TORCH_INTERNAL_ASSERT(
+        reinterpret_cast<size_t>(global_address) % 32 == 0,
+        "globalAddress, which specifies the starting address of the memory region described, must be 32 byte aligned when interleave is CU_TENSOR_MAP_INTERLEAVE_32B and 16 byte aligned otherwise.");
+  } else {
+    TORCH_INTERNAL_ASSERT(
+        reinterpret_cast<size_t>(global_address) % 16 == 0,
+        "globalAddress, which specifies the starting address of the memory region described, must be 32 byte aligned when interleave is CU_TENSOR_MAP_INTERLEAVE_32B and 16 byte aligned otherwise.");
+  }
+
+  for (auto global_dim_val : global_dim) {
+    constexpr cuuint64_t max_size = (cuuint64_t)1 << 32;
+    TORCH_INTERNAL_ASSERT(
+        global_dim_val != 0 && global_dim_val <= max_size,
+        "globalDim array, which specifies tensor size of each of the tensorRank dimensions, must be non-zero and less than or equal to 2^32.");
+  }
+
+  for (auto global_stride_val : global_strides) {
+    constexpr cuuint64_t max_stride = (cuuint64_t)1 << 40;
+    TORCH_INTERNAL_ASSERT(
+        global_stride_val % 16 == 0 && global_stride_val <= max_stride,
+        "globalStrides array, which specifies tensor stride of each of the lower tensorRank - 1 dimensions in bytes, must be a multiple of 16 and less than 2^40.");
+    if (interleave == CU_TENSOR_MAP_INTERLEAVE_32B) {
+      TORCH_INTERNAL_ASSERT(
+          global_stride_val % 32 == 0,
+          "The stride must be a multiple of 32 when interleave is CU_TENSOR_MAP_INTERLEAVE_32B.");
+    }
+  }
+  int64_t elem_size = (int64_t)dataTypeSize(dataType());
+  if (tensor_rank > 1) {
+    int64_t padding0 =
+        (int64_t)global_strides.at(0) - (int64_t)global_dim.at(0) * elem_size;
+    TORCH_INTERNAL_ASSERT(
+        padding0 >= 0,
+        "Negative pad0 for: globalStrides[0] = globalDim[0] * elementSizeInBytes(tensorDataType) + padding[0];");
+    for (int i = 1; i < (int64_t)tensor_rank - 1; i++) {
+      int64_t stride_mul_pad_i = (int64_t)global_strides.at(i) -
+          (int64_t)global_dim.at(i) * (int64_t)global_strides.at(i - 1);
+      TORCH_INTERNAL_ASSERT(
+          stride_mul_pad_i >= 0,
+          "Negative globalStrides[i – 1] * padding[i] for: globalStrides[i] = globalStrides[i – 1] * (globalDim[i] + padding[i]);");
+      // TODO: the check below is copied from the official doc, but does it
```

Comment:
I agree with you, this check is odd.

Reply:
Good catch! I will file an internal bug asking for clarification. For this PR, I will leave it as is (i.e., be consistent with the doc, right or wrong).